### ISSUE TITLE
Allow "filter" property for scheduling.userScheduler.plugins to add more scheduler plugins

### ISSUE THREAD

**Issue by a3626a** [jupyterhub/zero-to-jupyterhub-k8s#2808]
### Proposed change
`scheduling.userScheduler.plugins` configures kube-scheduler for better scheduling of singluser pods. kube-scheduler support various extension points like queueSort, preFilter, filter, postFilter, preScore, score, reserve, permit, preBind, bind, multiPoint. However Z2JH uses and allows only `score`.

Therefore, I suggest to allow other extension points. 

### Who would use this feature?

I am currently solving singleuser pod scheduling problem, related to this [issue](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/1851). So I implemented custom [kube-scheduler plugin](https://github.com/team-monolith-product/scheduler-plugins/blob/master/pkg/imagelocalityfilter/imagelocalityfilter.go) to filter out node without images. So users facing same problems might use this feature.

### (Optional): Suggest a solution

In this [configmap](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/main/jupyterhub/templates/scheduling/user-scheduler/configmap.yaml), `plugins` are inserted as yaml, so no additional Helm chart work is required. We just need to update schema.yaml.

We can simply allow `additionalProperties` for `scheduling.userScheduler.plugins` or add each extension points. After the implementation details are discussed, I can make PR for it.

**Comment by consideRatio** [comment]
Thanks for a great writeup @a3626a!!

Makes sense to relax the schema, allowing other properties than those explicitly listed (score) to be accepted.

I'd like us to not add more entries manually into this schema as its typically unsustainable to duplicate documentaton from k8s into this helm chart, so less is more in this case i think.

---

Btw, are you customizing what user scheduler image used to an image with your golang plugin compiled in the kube-scheduler binary, or can you compile a small file kube-scheduler can be configured to reference?

**Comment by a3626a** [comment]
> Btw, are you customizing what user scheduler image used to an image with your golang plugin compiled in the kube-scheduler binary, or can you compile a small file kube-scheduler can be configured to reference?

I am currently compiling entire user scheduler. so I replace the user-scheduler image into my custom one. I am not sure it is possible to compile single extension alone, and register it to vanilla kube-scheduler binary. It is my first time implementing kube-scheduler extension. I followed this [article](https://medium.com/@juliorenner123/k8s-creating-a-kube-scheduler-plugin-8a826c486a1).

```yaml
scheduling:
  userScheduler:
    image:
      name: <...>.dkr.ecr.ap-northeast-2.amazonaws.com/<...>-k8s-scheduler
      tag: <...>
```

**Comment by a3626a** [comment]
I will prepare PR for this.

**Comment by consideRatio** [comment]
Closed by #2809, thank you for so clearly describing intent and motivation etc!
